File: _Demo_and_Tests/Objective-C/AlgorithmsTests/BidirectionalBubbleSortTests.m
//
//  BidirectionalBubbleSortTests.m
//  Algorithms
//

#import <XCTest/XCTest.h>
#import "BidirectionalBubbleSort.h"
#import "TestsHelper.h"

@interface BidirectionalBubbleSortTests : XCTestCase

@end

@implementation BidirectionalBubbleSortTests

- (void)setUp {
    [super setUp];

    [TestsHelper reset];
}

- (void)tearDown {
    [super tearDown];
}

#pragma mark -

- (void)testAscOdd {
    [BidirectionalBubbleSort sort:TestsHelper.array order:BidirectionalBubbleSortOrderAsc];

    XCTAssertEqualObjects(TestsHelper.array, TestsHelper.sortedArrayAsc);
}

- (void)testAscEven {
    [TestsHelper.array addObject:@100];
    [TestsHelper.sortedArrayAsc addObject:@100];

    [BidirectionalBubbleSort sort:TestsHelper.array order:BidirectionalBubbleSortOrderAsc];

    XCTAssertEqualObjects(TestsHelper.array, TestsHelper.sortedArrayAsc);
}

- (void)testDescOdd {
    [BidirectionalBubbleSort sort:TestsHelper.array order:BidirectionalBubbleSortOrderDesc];

    XCTAssertEqualObjects(TestsHelper.array, TestsHelper.sortedArrayDesc);
}

- (void)testDescEven {
    [TestsHelper.array addObject:@100];
    [TestsHelper.sortedArrayDesc insertObject:@100 atIndex:0];

    [BidirectionalBubbleSort sort:TestsHelper.array order:BidirectionalBubbleSortOrderDesc];

    XCTAssertEqualObjects(TestsHelper.array, TestsHelper.sortedArrayDesc);
}

- (void)testNilArray {
    NSMutableArray *resultArray = nil;

    [BidirectionalBubbleSort sort:resultArray order:BidirectionalBubbleSortOrderAsc];

    XCTAssert(resultArray == nil);
}

- (void)testEmptyArray {
    NSMutableArray *resultArray = [NSMutableArray new];

    [BidirectionalBubbleSort sort:resultArray order:BidirectionalBubbleSortOrderAsc];

    XCTAssertEqualObjects(resultArray, @[]);
}

- (void)testArrayWithOneNumber {
    NSMutableArray *resultArray = [NSMutableArray arrayWithObject:@0];

    [BidirectionalBubbleSort sort:resultArray order:BidirectionalBubbleSortOrderAsc];

    XCTAssertEqualObjects(resultArray, @[@0]);
}

@end

File: _Demo_and_Tests/Swift/AlgorithmsTests/BidirectionalBubbleSortTests.swift
//
//  BidirectionalBubbleSortTests.swift
//  Algorithms
//

import XCTest

class BidirectionalBubbleSortTests: XCTestCase {

    override func setUp() {
        super.setUp()

        TestsHelper.clearArrays()
    }

    override func tearDown() {
        super.tearDown()
    }

    // MARK: -

    func testAscOdd() {
        let result = BidirectionalBubbleSort.sort(TestsHelper.array, order: .ASC)

        XCTAssert(result ?= TestsHelper.sortedArrayAsc)
    }

    func testAscEven() {
        TestsHelper.array.append(100)
        TestsHelper.sortedArrayAsc.append(100)

        let result = BidirectionalBubbleSort.sort(TestsHelper.array, order: .ASC)

        XCTAssert(result ?= TestsHelper.sortedArrayAsc)
    }

    func testDescOdd() {
        let result = BidirectionalBubbleSort.sort(TestsHelper.array, order: .DESC)

        XCTAssert(result ?= TestsHelper.sortedArrayDesc)
    }

    func testDescEven() {
        TestsHelper.array.append(100)
        TestsHelper.sortedArrayDesc.insert(100, at: 0)

        let result = BidirectionalBubbleSort.sort(TestsHelper.array, order: .DESC)

        XCTAssert(result ?= TestsHelper.sortedArrayDesc)
    }

    func testNilArray() {
        let nilArray: [Int]? = nil
        let result = BidirectionalBubbleSort.sort(nilArray, order: .ASC)

        XCTAssert(result == nil);
    }

    func testEmptyArray() {
        let result = BidirectionalBubbleSort.sort([] as [Int], order: .ASC)

        XCTAssert(result ?= []);
    }
    
    func testArrayWithOneNumber() {
        let result = BidirectionalBubbleSort.sort([0], order: .ASC)
        
        XCTAssert(result ?= [0]);
    }
    
}
